src/steps/containers.rs

```diff
@@ -0,0 +1,113 @@
+
+/// Returns a Vector of all containers, with Strings in the format
+/// "REGISTRY/[PATH/]CONTAINER_NAME:TAG"
+fn list_containers(crt: &Path) -> Result<Vec<String>> {
+    debug!(
+        "Querying '{} images --format \"{{{{.Repository}}}}:{{{{.Tag}}}}\"' for containers",
+        crt.display()
+    );
+    let output = Command::new(crt)
+        .args(&["images", "--format", "{{.Repository}}:{{.Tag}}"])
+        .output()?;
+    let output_str = String::from_utf8(output.stdout)?;
+
+    let mut retval = vec![];
+    for line in output_str.lines() {
+        if line.starts_with("localhost") {
+            // Don't know how to update self-built containers
+            debug!("Skipping self-built container '{}'", line);
+            continue;
+        }
+
+        if line.contains("<none>") {
+            // Bogus/dangling container or intermediate layer
+            debug!("Skipping bogus container '{}'", line);
+            continue;
+        }
+
+        debug!("Using container '{}'", line);
+        retval.push(String::from(line));
+    }
+
+    Ok(retval)
+}
+
+pub fn run_containers(ctx: &ExecutionContext) -> Result<()> {
+    // Prefer podman, fall back to docker if not present
+    let crt = match require("podman") {
+        Ok(path) => path,
+        Err(_) => require("docker")?,
+    };
+    debug!("Using container runtime '{}'", crt.display());
+
+    print_separator("Containers");
+    let mut success = true;
+    let containers = list_containers(&crt)?;
+    debug!("Containers to inspect: {:?}", containers);
+
+    for container in containers.iter() {
+        debug!("Pulling container '{}'", container);
+        let args = vec!["pull", &container[..]];
+        let mut exec = ctx.run_type().execute(&crt);
+
+        if let Err(e) = exec.args(&args).check_run() {
+            error!("Pulling container '{}' failed: {}", container, e);
+
+            // Find out if this is 'skippable'
+            // This is necessary e.g. for docker, because unlike podman docker doesn't tell from
+            // which repository a container originates (such as `docker.io`). This has the
+            // practical consequence that all containers, whether self-built, created by
+            // docker-compose or pulled from the docker hub, look exactly the same to us. We can
+            // only find out what went wrong by manually parsing the output of the command...
+            if match exec.check_output() {
+                Ok(s) => s.contains(NONEXISTENT_REPO),
+                Err(e) => match e.downcast_ref::<TopgradeError>() {
+                    Some(TopgradeError::ProcessFailedWithOutput(_, stderr)) => {
+                        if stderr.contains(NONEXISTENT_REPO) {
+                            true
+                        } else {
+                            return Err(e);
+                        }
+                    }
+                    _ => return Err(e),
+                },
+            } {
```

Comment:
I think it can be made much simpler if no error will cause the function to quit. This is similar to what we do when pulling git repositories. If any repo fails updating then `success` will be made `false` and the step will eventually be marked as failed, but I don't think we should stop trying the rest of the repos in any case.